### DISCUSSION TITLE
Training with a public Kaggle Notebook

### DISCUSSION THREAD

**Discussion by Green0-0** [VsonicV/es-fine-tuning-paper#37] [Ideas]
I've noticed that Kaggle gives everyone free access to a TPUv5e-8 for 20 hours a week.
I benchmarked the running speeds, and this device has a 20% faster generation/prompt process speed than a single 5090 (around ~4k t/s with an 8b model).
I have also succeeded in perturbing the model and restoring it from the same seed.
I'm currently working on something else, but if anyone could finish the notebook and push it to this repo, that would allow anyone to replicate the ES fine tuning experiments online through Kaggle and thus remove the hardware barrier.

For reference, here is the code:
```
!pip install uv
!uv pip install vllm-tpu --system 
```
```
import os
import sys
import io
import jax
import jax.numpy as jnp
from flax import nnx
from vllm import LLM, SamplingParams

# Set environment variables for single device
os.environ.pop("TPU_MULTIHOST_BACKEND", None)
# Disable V1 multiprocessing to run engine in-process
os.environ["VLLM_ENABLE_V1_MULTIPROCESSING"] = "0"

print("Initializing LLM on single TPU device...")
model_name = "Qwen/Qwen3-4B"

llm = LLM(
    model=model_name,
    tensor_parallel_size=8,
    trust_remote_code=True,
    dtype="bfloat16",
    max_model_len=4096,
    gpu_memory_utilization=0.6, # Leave room for perturbation tensors
)
```
```
import logging
import gc
logging.getLogger("vllm.tpu_inference").setLevel(logging.WARNING)


prompts = ["Hello my name is"]
# Deterministic generation for demo purposes
sampling_params = SamplingParams(temperature=0.0, max_tokens=20)

print("-" * 50)
print("1. Baseline Generation")
outputs = llm.generate(prompts, sampling_params)
print(f"Output: {outputs[0].outputs[0].text!r}")

# Access the worker and model runner directly
# With VLLM_ENABLE_V1_MULTIPROCESSING=0, we should be able to navigate to the worker
print("\nAccessing model worker...")

# Debug helper to find the path
if hasattr(llm.llm_engine, 'model_executor'):
    worker = llm.llm_engine.model_executor.driver_worker
elif hasattr(llm.llm_engine, 'engine') and hasattr(llm.llm_engine.engine, 'model_executor'):
      worker = llm.llm_engine.engine.model_executor.driver_worker
else:
    # Fallback: inspect structure
    print(f"DEBUG: Engine type: {type(llm.llm_engine)}")
    print(f"DEBUG: Engine dir: {dir(llm.llm_engine)}")
    raise AttributeError("Could not find model_executor even with MP disabled.")

model_runner = worker.model_runner

# Capture original state
print("Capturing original state...")
original_state = model_runner.state

# Helper function to process and sync weights in chunks
def process_and_sync_in_chunks(worker, seed, std, operation="add"):
    print(f"Starting chunked {operation} operation...")
    
    # Always get the latest state from the runner
    # This is crucial because sync_weights updates the state object reference
    state = worker.model_runner.state
    flat_state = list(state.flat_state())
    # Sort by path to ensure deterministic order regardless of state changes
    flat_state.sort(key=lambda x: str(x[0]))
    
    total_params = len(flat_state)
    chunk_size = 10 
    
    # Initialize PRNG key
    key = jax.random.PRNGKey(seed)
    
    for i in range(0, total_params, chunk_size):
        # Re-fetch state for each chunk to ensure we are working with the latest values
        # (Although for distinct chunks, the previous chunk's update shouldn't affect this one's values,
        #  it ensures we have the valid object reference)
        current_state = worker.model_runner.state
        # We need to reconstruct the flat list or access by path?
        # Re-flattening every time is expensive.
        # But since we iterate 'flat_state' which is a list of (path, value),
        # the 'value' here might be STALE if 'state' became stale.
        # However, we only need the PATH to find the value in the current state.
        
        # Actually, let's stick to the plan:
        # We iterate the INITIAL flat_state structure to get paths and keys.
        # But we must read the VALUE from the CURRENT state.
        
        chunk_paths = [item[0] for item in flat_state[i : i + chunk_size]]
        
        chunk_update = {}
        chunk_mappings = {}
        
        # Helper to find value in current state by path
        def get_value_by_path(root, path):
            node = root
            for p in path:
                node = node[p]
            return node

        for path in chunk_paths:
            # Get current value from the LIVE model state
            val = get_value_by_path(current_state, path)
            
            # Extract value and sharding
            leaf = val
            sharding = None
            
            if hasattr(val, 'value'):
                leaf = val.value
            
            if hasattr(val, 'sharding'):
                sharding = val.sharding
            elif hasattr(leaf, 'sharding'):
                sharding = leaf.sharding
            
            # Generate noise
            if isinstance(leaf, jax.Array) and jnp.issubdtype(leaf.dtype, jnp.floating):
                key, subkey = jax.random.split(key)
                noise = jax.random.normal(subkey, leaf.shape, dtype=leaf.dtype) * std
                
                # Debug: Check noise consistency
                if i == 0 and path == chunk_paths[0]:
                    print(f"DEBUG: First noise mean: {jnp.mean(noise)}")
                    print(f"DEBUG: First leaf mean: {jnp.mean(leaf)}")
                
                if operation == "add":
                    new_val = leaf + noise
                elif operation == "sub":
                    new_val = leaf - noise
                else:
                    new_val = leaf
            else:
                new_val = leaf
            
            key_str = '.'.join(str(k) for k in path)
            
            class SimpleParam:
                def __init__(self, value):
                    self.value = value
            
            chunk_update[key_str] = SimpleParam(new_val)
            
            if sharding is not None:
                chunk_mappings[key_str] = (key_str, sharding)
        
        if chunk_update:
            chunk_state = nnx.State(chunk_update)
            worker.sync_weights(
                updated_weights=chunk_state,
                mappings=chunk_mappings,
                transpose_keys={},
                reshard_fn=None
            )
            
            arrays_to_sync = [p.value for p in chunk_update.values()]
            jax.block_until_ready(arrays_to_sync)
        
        del chunk_update
        del chunk_mappings
        
        gc.collect()
        
        print(f"Processed chunk {i//chunk_size + 1}/{(total_params + chunk_size - 1)//chunk_size}", end='\r')
    print("\nDone.")

# Generate perturbation
print("Generating and syncing perturbation (Seed 42, Std 0.1)...")
process_and_sync_in_chunks(worker, seed=42, std=0.1, operation="add")

print("-" * 50)
print("2. Perturbed Generation")
outputs = llm.generate(prompts, sampling_params)
print(f"Output: {outputs[0].outputs[0].text!r}")

# Restore Original Weights
print("\nRestoring original weights (Subtraction)...")
process_and_sync_in_chunks(worker, seed=42, std=0.1, operation="sub")

print("-" * 50)
print("3. Restored Generation")
outputs = llm.generate(prompts, sampling_params)
print(f"Output: {outputs[0].outputs[0].text!r}")
```
Which prints
```
--------------------------------------------------
1. Baseline Generation

Adding requests: 100%
 1/1 [00:00<00:00, 78.55it/s]
Processed prompts: 100%
 1/1 [00:00<00:00,  6.82it/s, est. speed input: 27.27 toks/s, output: 136.33 toks/s]

Output: " Alex, I'm a 22-year-old male, and I'm a student at the University"

Accessing model worker...
Capturing original state...
Generating and syncing perturbation (Seed 42, Std 0.1)...
Starting chunked add operation...
DEBUG: First noise mean: -0.00123596
DEBUG: First leaf mean: 2.93255e-05
Processed chunk 40/40
Done.
--------------------------------------------------
2. Perturbed Generation

Adding requests: 100%
 1/1 [00:00<00:00, 149.34it/s]
Processed prompts: 100%
 1/1 [00:01<00:00,  1.30s/it, est. speed input: 3.07 toks/s, output: 15.33 toks/s]

Output: ' Danielscele另有\':[\'VENT.recycle ============================================================================\ntetCloud将于 Selected#define사를尺度 };\n\n\n\'Lşık"profile virtualความ'

Restoring original weights (Subtraction)...
Starting chunked sub operation...
DEBUG: First noise mean: -0.00123596
DEBUG: First leaf mean: -0.00120544
Processed chunk 40/40
Done.
--------------------------------------------------
3. Restored Generation

Adding requests: 100%
 1/1 [00:00<00:00, 152.25it/s]
Processed prompts: 100%
 1/1 [00:00<00:00,  6.42it/s, est. speed input: 25.67 toks/s, output: 128.32 toks/s]

Output: " Alex, I'm a 22-year-old student. I'm from the UK, and I"
```

Note: the above code runs in tensor parallel across the 8 tpu devices, but it would technically be faster to run the model separately per device.

**Comment by VsonicV** [comment]
It is a good suggestion! Thanks for sharing the code! We haven't tested the implementation on TPU, but would love to try.
